extension/shops/pilulka.mjs

```diff
@@ -1,24 +1,56 @@
-import { cleanPrice, registerShop } from "../helpers.mjs";
-import { AsyncShop } from "./shop.mjs";
+import { cleanPrice, registerShop, getLdJsonByType } from "../helpers.mjs";
+import { StatefulShop } from "./shop.mjs";
+
+export class Pilulka extends StatefulShop {
 
-export class Pilulka extends AsyncShop {
   get injectionPoint() {
     return ["afterend", ".service-detail__basket-box"];
   }
 
-  get waitForSelector() {
-    return ".service-detail__basket-box";
+  // detailSelector is used during scheduleRendering for initial widget render,
+  // but we wanna skip that and only render when shouldRender is called from MutationObserver
+  // Note that removing detailSelector would trigger an error so we return dummy value
+  get detailSelector() {
+    return "nonsense-value-to-never-render-during-initial-scheduleRendering";
+  }
+
+  shouldRender(mutations) {
+    // rating box is loaded last
+    // = safest time to render widget without worrying about it being removed
+    const needle = mutations.find(mutation =>
+      Array.from(mutation.addedNodes).find(node =>
+        node.querySelector?.(".service-detail__main__right__inner .rating--box")
+      )
+    )
+    // console.log(`shouldRender: ${needle ? '✅' : '🔴'}`);
+    return !!needle;
+  }
+
+  shouldCleanup(mutations) {
+    const needle = mutations.find(mutation =>
+      Array.from(mutation.removedNodes).find(node =>
+        node.getAttribute?.("componentname") === "catalog.product"
+      )
+    )
+    // console.log(`shouldCleanup: ${needle ? '✅' : '🔴'}`);
+    return !!needle;
   }
 
+
   async scrape() {
-    const data = JSON.parse(document.querySelector("script[type='application/ld+json']").textContent);
-    const product = data.find(x => x["@type"] === "Product");
+    // Give Nuxt 100ms to replace the json-ld script tag with correct data,
+    // otherwise it could contain old product when using SPA navigation
+    await new Promise(resolve => setTimeout(resolve, 100));
```

Comment:
Please, wait for the observed change of element. Don't introduce arbitrary race conditions.

Reply:
@rarous  @Strajk Can we revive this thread and complete the fix? I tried to fix it in https://github.com/topmonks/hlidac-shopu/commit/e8486d0b781b99f051e59df254089a02c3521181 unaware of this existing PR, but obviously, the extension still fails to render with my solution in some cases...